File: mergeNsortedLists.cpp
/********************************************************
*       Merge N sorted lists
*       input format:
*           first string - number of input lists
*           next n strings - the next list (first number in string - length of list)
*			maximum value in list should be 100
*       output format:
*           string with merged lists
*********************************************************/

#include "stdafx.h"
#include <algorithm>
#include <string>
#include <iostream>
#include <vector>

using std::string;
using std::vector;
using std::cin;
using std::cout;
using std::getline;
using std::cout;
using std::min_element;
using std::to_string;

int main() {

    string lengthS;
    getline(cin, lengthS);
    int length = stoi(lengthS);
    if (length == 0)
        return 0;
    int count = 0;
    string inputStr;
    string digit;
    vector<vector<short int>> InputLists;
    vector<vector<short int>::iterator> InputListsIt;
    int totalLen = 0;
    bool getFirstDig = true;
    while (count < length) {
        InputLists.push_back({});
        getline(cin, inputStr);

        digit = "";
        getFirstDig = true;
        for (int i = 0; i < inputStr.length(); i++) {
            if (inputStr[i] == ' ') {
                InputLists[count].push_back(stoi(digit));
                if (getFirstDig) {
                    totalLen += stoi(digit);
                    getFirstDig = false;
                }
                digit = "";
            }
            digit += inputStr[i];
        }
        InputLists[count].push_back(stoi(digit)); // push last digit
        InputListsIt.push_back(InputLists[count].begin() + 1);

        count++;
    }

    string result = "";
    vector<int> compare;
    vector<int>::iterator minCompare;
    bool flag = true;
    int min = 101; // maximal possible value 100
    int idx;
    int len = 0;

    while (len != totalLen) {
        for (int i = 0; i < InputLists.size(); i++) {
            if (InputListsIt[i] != InputLists[i].end())
                compare.push_back(*InputListsIt[i]);
            else
                compare.push_back(min);
        }

        minCompare = min_element(begin(compare), end(compare));
        auto it = begin(compare);
        while ((it = find(it, end(compare), *minCompare)) != end(compare)) {
            idx = distance(begin(compare), it++);
            if (InputListsIt[idx] != InputLists[idx].end()) {
                InputListsIt[idx] += 1;
                result += to_string(*minCompare) + ' ';
                len++;
            }
        }
        compare.clear();
    }

    cout << result << std::endl;
    return 0;
}

File: mergeNsortedLists.py
##-----------------------------------------------------------##
##       Merge N sorted lists
##       input format:
##           first string - number of input lists
##           next n strings - the next list (first number in string - length of list)
##			maximum value in list should be 100
##       output format:
##           string with merged lists
##-----------------------------------------------------------##
import sys

length = int(sys.stdin.readline().strip())
inputLists = []
MAXIMAL_VALUE = 101
i = 0
if length == 0:
    pass
else:
	while i < length:
		inputLists.append(list(map(lambda x: int(x), sys.stdin.readline().strip().split(' '))))
		i += 1

ans = ''

def findMinK(findMinArr):
    min = MAXIMAL_VALUE
    idx = 0
    idxMin = []
    for v in findMinArr:
        if (v == min):
            idxMin.append(idx)
        if v < min:
            idxMin = []
            min = v
            idxMin.append(idx)
        idx += 1

    return [idxMin, min]


idxs = []
lens = []
totalLen = 0
for i in range(length):
    lens.append(inputLists[i][0])
    totalLen += inputLists[i][0]
    idxs.append(1) if inputLists[i][0] != 0 else idxs.append(-1)

lenAns = 0
while lenAns < totalLen:
    findMinArr = []
    cnt_idx = 0
    for idx in idxs:
        if idx != -1:
            findMinArr.append(inputLists[cnt_idx][idx])
        else:
            findMinArr.append(MAXIMAL_VALUE)
        cnt_idx += 1

    [idxMins, min] = findMinK(findMinArr)

    for idxMin in idxMins:
        ans += str(min) + ' '
        lenAns += 1

        flag = 1
        while flag:
            if idxs[idxMin] != lens[idxMin]:
                if inputLists[idxMin][idxs[idxMin] + 1] == min:
                    ans += str(inputLists[idxMin][idxs[idxMin] + 1]) + ' '
                    lenAns += 1
                    idxs[idxMin] += 1
                else:
                    flag = 0
            else:
                idxs[idxMin] = -1
                flag = 0

        if idxs[idxMin] != -1:
            idxs[idxMin] += 1
        if idxs[idxMin] == lens[idxMin] + 1:
            idxs[idxMin] = -1

print(ans)